beaker/application.py

```diff
@@ -1,418 +1,963 @@
+                    raise ValueError(
+                        f"override=False, but bare external for {for_action} already exists."
                     )
+                assert isinstance(existing_action.action, SubroutineFnWrapper)
+                self.deregister_bare_method(existing_action.action)
+            self.bare_actions[for_action] = OnCompleteAction(
+                action=sub, call_config=call_config
+            )
 
-        self.on_create: Optional[HandlerFunc] = (
-            all_creates.pop() if len(all_creates) == 1 else None
-        )
-        self.on_update: Optional[HandlerFunc] = (
-            all_updates.pop() if len(all_updates) == 1 else None
+    def deregister_bare_method(
+        self,
+        action_name_or_reference: OnCompleteActionName | SubroutineFnWrapper,
+        /,
+    ) -> None:
+        if isinstance(action_name_or_reference, str):
+            self.bare_actions.pop(cast(OnCompleteActionName, action_name_or_reference))
+        else:
+            method = action_name_or_reference
+            _remove_first_match(self.bare_actions, lambda _, v: v.action is method)
+
+    @overload
+    def external(
+        self,
+        fn: HandlerFunc,
+        /,
+    ) -> ABIReturnSubroutine:
+        ...
+
+    @overload
+    def external(
+        self,
+        /,
+        *,
+        method_config: MethodConfig | MethodConfigDict | None = None,
+        name: str | None = None,
+        authorize: SubroutineFnWrapper | None = None,
+        bare: bool | None = False,
+        read_only: bool = False,
+        override: bool | None = False,
+    ) -> DecoratorFuncType:
+        ...
+
+    def external(
+        self,
+        fn: HandlerFunc | None = None,
+        /,
+        *,
+        method_config: MethodConfig | MethodConfigDict | None = None,
+        name: str | None = None,
+        authorize: SubroutineFnWrapper | None = None,
+        bare: bool | None = False,
+        read_only: bool = False,
+        override: bool | None = False,
+    ) -> DecoratorResultType | DecoratorFuncType:
+        """
+        Add the method decorated to be handled as an ABI method for the Application
+
+        Args:
+            fn: The function being wrapped.
+            method_config:  <TODO>
+            name: Name of ABI method. If not set, name of the python method will be used.
+                Useful for method overriding.
+            authorize: a subroutine with input of ``Txn.sender()`` and output uint64
+                interpreted as allowed if the output>0.
+            bare:
+            read_only: Mark a method as callable with no fee using dryrun or simulate
+            override:
+
+        Returns:
+            The original method with additional elements set in it's
+            :code:`__handler_config__` attribute
+        """
+
+        def decorator(func: HandlerFunc) -> SubroutineFnWrapper | ABIReturnSubroutine:
+            sig = inspect.signature(func)
+            nonlocal bare
```

Comment:
had to google this, somehow never seen it before

Reply:
probably a good thing, it's usually a code smell. in this case you can't both assign to it and read from it, since it's in a closure. And I didn't know what else to call the new variable - I'll remove this smell and just call it `bare_`